### DISCUSSION TITLE
session.execute for a stored procedure

### DISCUSSION THREAD

**Discussion by renatocfrancisco** [sqlalchemy/sqlalchemy#9807] [Usage Questions]
```python
Session = sessionmaker(bind=engine)
session = Session()
session.begin()

date = '23/05/2023'
value = 1

sql = text("""exec dbo.procedure :v1, :v2""")
params = {
    "v1": value,
    "v2": date
}
session.execute(sql, params)
```

I am using sqlalchemy v.2.0.14 for sql server.
I tested with `Session(engine)`, I tested inserting the params object after the text, doesn't work.
Now, this code sends an error: `expected a simple type, a tuple or a list`.

The engine is correct. The user has access to execute procedures. I check the documentation, i am so confused.
Basically, how do you execute a procedure with pure sql in sqlalchemy?

**Comment by zzzeek** [comment]
assuming this is pyodbc, pyodbc's syntaxes for stored procs are in their documentation at 
https://github.com/mkleehammer/pyodbc/wiki/Calling-Stored-Procedures

given a Session you might want to use [Session.connection()](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.connection) then [Connection.exec_driver_sql()](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Connection.exec_driver_sql)

```py

session.connection().exec_driver_sql("{CALL usp_NoParameters}")
```


the DBAPI also has a callproc interface.   We dont supply this directly, if you need that you would use a [raw cursor](https://docs.sqlalchemy.org/en/20/core/connections.html#calling-stored-procedures-and-user-defined-functions).

**Reply by renatocfrancisco** [reply]
oh, yeah i forgot. I am using pymssql instead of pyodbc.

**Reply by zzzeek** [reply]
OK here's their docs: https://pymssql.readthedocs.io/en/latest/pymssql_examples.html?highlight=stored#calling-stored-procedures

you'll want to use callproc


```py
raw_conn = session.connection().raw_connection()
cursor = raw_conn.cursor()
cursor.callproc(...)
```

**Comment by renatocfrancisco** [comment]
i am so sorry @zzzeek . i managed to solve the problem, but without using the pymssql doc.
Apparently this worked (using v2.0.12):
```python
query = f"exec dbo.procedure {value}, '{date}'"

session = Session(engine)
session.execute(text(query))
```

🤷🏻
